libsolidity/parsing/Parser.cpp

```diff
@@ -81,13 +81,14 @@ class Parser::ASTNodeFactory
 	SourceLocation m_location;
 };
 
-ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
+ASTPointer<SourceUnit> Parser::parse(CharStream& _charStream)
 {
 	solAssert(!m_insideModifier, "");
 	try
 	{
 		m_recursionDepth = 0;
-		m_scanner = _scanner;
+		m_source = &_charStream;
+		m_scanner = make_shared<Scanner>(_charStream);
```

Comment:
Does this really need a shared pointer?

Reply:
Oh it kind of does. A `unique_ptr` would probably be better, but for that you need the full Scanner in the header. Although scanner.h is not that big anymore, is it? Maybe `std::optional<Scanner>` could also work.

Reply:
Oh I am all in favor of preferring `unique_ptr`, it's just that the whole repo is full of `shared_ptr`'s. I"ll adapt. :-)

Reply:
tried and reverted as it's only working if `Scanner` is fully declared for use with `unique_ptr` ;-(